uast/nodes/nodesproto/nodesproto.go

```diff
@@ -296,6 +298,54 @@ func ReadTree(r io.Reader) (nodes.Node, error) {
 	return g.asTree()
 }
 
+type RawNode struct {
+	ID     uint64      `json:"id"`
+	Kind   nodes.Kind  `json:"kind"`
+	Value  nodes.Value `json:"val,omitempty"`
+	Keys   []uint64    `json:"keys,omitempty"`
+	Values []uint64    `json:"values,omitempty"`
+}
+
+type RawGraph struct {
+	Root  uint64    `json:"root,omitempty"`
+	Meta  uint64    `json:"meta,omitempty"`
+	Last  uint64    `json:"last,omitempty"`
+	Nodes []RawNode `json:"nodes"`
```

Comment:
any reason to have it as an array instead of the map?

RawNode keeps ids in `Keys` and `Values` but those are original ids.
Most probably to read them I can do
```go
for _, id := range n.Values {
  fmt.Println(gr.Nodes[id - gr.Root])
}
```
but it feels unnatural and error prune. 

or have I misunderstood something?

Reply:
You don't need to subtract root, just make a lookup by id.

Yeah, I was thinking about a map, but it will be unordered in this case.

Reply:
Hm. Sorry, I must missing something.

But the default id of root is 1 right? And this is an array. So `gr.Nodes[gr.Root]` would return me the next node after root, right? And the same for all other ids. The would have `id - 1` index in the array. That's why I subtract root.

Even in tests I see a node has `id=1` but stays on `0` element of the array,

Reply:
First, nodes are sparse, meaning that you can have nodes `1,7,13`, and array indexes will be `0,1,2`, so you can't use an ID as an array index and you cannot calculate indexes. The only guarantee is that nodes are ordered by ID. 

Second, RootID can be any valid node ID from that array. For example it can be `7`.

You see root as a node with `id=1` is only because encoder emits nodes in BFS order. But this order can be arbitrary.

Reply:
I will change it to a map, then. Requiring sorting is definitely less common.